packages/runtime/container-runtime/src/test/documentSchema.spec.ts

```diff
@@ -97,6 +99,102 @@ describe("Runtime", () => {
 		testWrongConfig({ ...validConfig, runtime: { ...validConfig.runtime, foo: 5 } });
 	});
 
+	it("disallowed versions", () => {
+		const controller = createController({
+			...validConfig,
+			runtime: { ...validConfig.runtime, disallowedVersions: [] },
+		});
+		assert(controller.sessionSchema.runtime.disallowedVersions === undefined);
+		controller.onMessageSent(() => {
+			assert(false, "no messages should be sent!");
+		});
+
+		createController({
+			...validConfig,
+			runtime: { ...validConfig.runtime, disallowedVersions: ["aaa"] },
+		});
+		testWrongConfig({
+			...validConfig,
+			runtime: { ...validConfig.runtime, disallowedVersions: [pkgVersion] },
+		});
+		testWrongConfig({
+			...validConfig,
+			runtime: { ...validConfig.runtime, disallowedVersions: ["aaa", pkgVersion, "bbb"] },
+		});
+	});
+
+	it("change disallowed versions", () => {
+		const controller = new DocumentsSchemaController(
+			true, // existing,
+			// old schema
+			{
+				...validConfig,
+				runtime: { ...validConfig.runtime, explicitSchemaControl: true },
+			},
+			// features requested
+			{
+				...features,
+				disallowedVersions: ["aaa", "bbb"],
+			},
+			// onSchemaChange
+			() => {},
+		);
+		assert.deepEqual(controller.sessionSchema.runtime.disallowedVersions, ["aaa", "bbb"]);
+		let message: IDocumentSchemaCurrent | undefined;
+		controller.onMessageSent((msg) => {
+			message = msg as IDocumentSchemaCurrent;
+		});
+		assert(message !== undefined);
+		controller.processDocumentSchemaOp(message, true /* local */, 100 /* sequenceNumber */);
```

Comment:
I am not sure I understand what's the use of sending this message. Why would this change the disallowedVersions?

Reply:
If this op contained another disallowed version, say "ddd", would the schema for this controller change to include "ddd"?

Reply:
This code validates op roundtrip flow. Controller can generate an op, and if it does, eventually it routdrips and gets processed by controller - that's what this code emulates (emulating end-to-end flow, including interaction with container runtime).

I've added some comments on functions themselves - please glance.

Reply:
It makes more sense now. There are a couple of things that are confusing though:
1. controller.onMessageSent() above updates the `message` with the controller's document schema. This is not obvious at all. Maybe adding a comment or a helper that says this is what is happening may be helpful. Or maybe that function should be renamed to something like `onMessageSent` -> `maybeSendSchemaChangeMessage`.
2. The schema change op here is basically a no-op. It would be interesting to add a scenario where one controller sends an op with additional disallowed versions and these are adde to another containers list.

Reply:
(Adding this same text to the header comment above class)

Controller operates with 4 schemas:
- document schema: whatever we loaded from summary metadata + ops
- desired schema - what client is asking for to have (i.e. all the desired settings, based on runtime options / feature gates)
- session schema - current session schema. It's "and" of the above two schemas
- future schema - "or" of document and desires schemas.

"or" & "and" operators are defined individually for each property. For Boolean properties it's literally &&, || operators.
But for other properties it's more nuanced.

The op is sent because document schema does not match future schema. Op changes documents schema to future schema.
The fact that future schema is the same as session schema is somewhat irrelevant.
And no, it's not noop. It's somewhat noop for this session, yes, but it's not noop for all the other sessions who might find out that they are not welcome in this document.